### PR TITLE
#1465: Use external synchronization of items instead of java locks.

### PR DESCRIPTION
**Pull request by paulodamaso**
For #1465:
- Created `com.zerocracy.farm.sync.Lock` interface for modeling locking behaviors for Zerocracy 
- Replaced `java.util.concurrent.locks.Lock` usages in `SmartLock` for `com.zerocracy.farm.sync.Lock`
- Created `MongoDBLock` skeleton for lock implementation in database
- Created tests for `MongoDBLock`
- Left puzzle for `MongoDBLock` implementation

### CODE REVIEW

**Comment by 0crat** [comment]
Job [#1630](https://github.com/zerocracy/farm/issues/1630) is now in scope, role is `REV`

File: src/main/java/com/zerocracy/farm/sync/MongoDBLock.java [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+package com.zerocracy.farm.sync;
+
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.Condition;
+
+/**
+ * {@link Lock} using MongoDB.
+ *
+ * @since 1.0
+ * @todo #1465:30min Implement external locking mechanism based in MongoDB.
```
Comment (krzyk) [inline src/main/java/com/zerocracy/farm/sync/MongoDBLock.java]: @paulodamaso you didn't remove the origin `todo`

File: src/main/java/com/zerocracy/farm/sync/MongoDBLock.java [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+ *  The lock must be set to some resource in a table where resource path and
+ *  name must be set, along with time of lock. Then remove expects from
+ *  MongoDBLockTest.
+ * @checkstyle AbbreviationAsWordInNameCheck (500 lines)
+ */
+public final class MongoDBLock implements Lock {
```
Comment (krzyk) [inline src/main/java/com/zerocracy/farm/sync/MongoDBLock.java]: @paulodamaso I would change `DB` to `Db` and get rid of checkstyle exclude

File: src/test/java/com/zerocracy/farm/sync/MongoDBLockTest.java [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+import org.hamcrest.core.IsNot;
+import org.hamcrest.text.IsEmptyString;
+import org.junit.Test;
+
+/**
+ * Test case for {@link MongoDBLock}.
+ *
+ * @since 1.0
+ * @checkstyle JavadocMethodCheck (500 lines)
+ * @checkstyle AbbreviationAsWordInNameCheck (500 lines)
```
Comment (krzyk) [inline src/test/java/com/zerocracy/farm/sync/MongoDBLockTest.java]: @paulodamaso and here also, you can remove checkstyle exclude by changing the class name

File: src/main/java/com/zerocracy/farm/sync/MongoDBLock.java (hunk 1 above)
Comment (paulodamaso) [reply src/main/java/com/zerocracy/farm/sync/MongoDBLock.java]: @krzyk Actually I've updated the puzzle for using MongoDB as the locking implementation and added some implementation specifications.

Comment (paulodamaso) [reply src/main/java/com/zerocracy/farm/sync/MongoDBLock.java]: @krzyk Actually I've updated the puzzle for using MongoDB as the locking implementation and added some implementation specifications.

Comment (krzyk) [reply src/main/java/com/zerocracy/farm/sync/MongoDBLock.java]: @paulodamaso but you need to remove the original todo.

**Comment by paulodamaso** [comment]
@krzyk thanks, I forgot about the puzzle. Corrected, please take a look.

**Comment by krzyk** [comment]
@paulodamaso thanks

**Comment by krzyk** [comment]
@rultor merge

**Comment by paulodamaso** [comment]
@g4s8 ping

**Comment by g4s8** [comment]
@rultor merge

**Comment by 0crat** [comment]
Job was finished in 38 hours, bonus for fast delivery is possible (see [§36](http://www.zerocracy.com/policy.html#36))

**Comment by 0crat** [comment]
@ypshenychka[/z](https://www.0crat.com/u/ypshenychka) please review this job completed by @krzyk[/z](https://www.0crat.com/u/krzyk), as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

**Comment by 0crat** [comment]
The job [#1630](https://github.com/zerocracy/farm/issues/1630) is now out of scope

**Comment by 0crat** [comment]
Payment to `ARC` for a closed pull request, as in [§28](http://www.zerocracy.com/policy.html#28): +10 point(s) just awarded to @g4s8[/z](https://www.0crat.com/u/g4s8)

**Comment by ypshenychka** [comment]
@krzyk According to our [QA Rules](http://www.zerocracy.com/policy.html#42):
>The code reviewer found at least three problems in the code.
Comments were mostly about design problems, not cosmetic issues.

Only one major issue was found during code review. 
Please confirm that you'll try to find at least three major problems while future reviews.

**Comment by krzyk** [comment]
@ypshenychka I confirm

**Comment by ypshenychka** [comment]
@krzyk thanks

**Comment by ypshenychka** [comment]
@0crat quality acceptable

**Comment by 0crat** [comment]
Order was finished, quality is "acceptable": +20 point(s) just awarded to @krzyk[/z](https://www.0crat.com/u/krzyk)

**Comment by 0crat** [comment]
Quality review completed: +8 point(s) just awarded to @ypshenychka[/z](https://www.0crat.com/u/ypshenychka)

**Pull request merged**